### PR TITLE
validation tip added for email colors setting ref #17168

### PR DESCRIPTION
**Pull request by nishitlangaliya**
Fixes.
JQuery function added for check pattern of hexa color code. please check for the same and let me know if any changes required.

Thanks

### CODE REVIEW

File: assets/js/admin/woocommerce_admin.js [hunk 1]
```diff
@@ -296,4 +296,14 @@ jQuery( function ( $ ) {
 			$( this ).replaceWith( '<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/' + video_id + '" frameborder="0" allowfullscreen></iframe>' );
 		} );
 	});
+
+	// Validate email colors settings
+	$( '#woocommerce_email_base_color[type=text], #woocommerce_email_background_color[type=text], #woocommerce_email_body_background_color[type=text], #woocommerce_email_text_color[type=text]' ).on( 'change', function() {
+		var hexcolor = jQuery(this).val();
+		if (! hexcolor.match(/^\#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$/)) {
+			jQuery( document.body ).triggerHandler( 'wc_add_error_tip', [ jQuery(this), 'hexa_color_error' ] );
```
Comment (kloon) [inline assets/js/admin/woocommerce_admin.js]: For consistency, please use $ here instead of jQuery

File: assets/js/admin/woocommerce_admin.js [hunk 2]
```diff
@@ -296,4 +296,14 @@ jQuery( function ( $ ) {
...
+        } else {
+			jQuery( document.body ).triggerHandler( 'wc_remove_error_tip', [ jQuery(this), 'hexa_color_error' ] );
```
Comment (kloon) [inline assets/js/admin/woocommerce_admin.js]: For consistency, please use $ here instead of jQuery

File: assets/js/admin/woocommerce_admin.js [hunk 3]
```diff
@@ -296,4 +296,14 @@ jQuery( function ( $ ) {
 			$( this ).replaceWith( '<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/' + video_id + '" frameborder="0" allowfullscreen></iframe>' );
 		} );
 	});
+
+	// Validate email colors settings
+	$( '#woocommerce_email_base_color[type=text], #woocommerce_email_background_color[type=text], #woocommerce_email_body_background_color[type=text], #woocommerce_email_text_color[type=text]' ).on( 'change', function() {
+		var hexcolor = jQuery(this).val();
+		if (! hexcolor.match(/^\#([a-fA-F0-9]{6}|[a-fA-F0-9]{3})$/)) {
```
Comment (kloon) [inline assets/js/admin/woocommerce_admin.js]: Space before `!` and spaces around `this`

File: assets/js/admin/woocommerce_admin.js [hunk 4]
```diff
@@ -296,4 +296,14 @@ jQuery( function ( $ ) {
 			$( this ).replaceWith( '<iframe width="560" height="315" src="https://www.youtube-nocookie.com/embed/' + video_id + '" frameborder="0" allowfullscreen></iframe>' );
 		} );
 	});
+
+	// Validate email colors settings
+	$( '#woocommerce_email_base_color[type=text], #woocommerce_email_background_color[type=text], #woocommerce_email_body_background_color[type=text], #woocommerce_email_text_color[type=text]' ).on( 'change', function() {
+		var hexcolor = jQuery(this).val();
```
Comment (kloon) [inline assets/js/admin/woocommerce_admin.js]: Spaces around `this`

File: assets/js/admin/woocommerce_admin.js (hunk 3 above)
Comment (kloon) [inline assets/js/admin/woocommerce_admin.js]: Check your spaces, coding standards say there needs to be spaces before and after brackets.

**Review by kloon — CHANGES_REQUESTED** [review]
Happy with this once the coding standards are addressed.

**Comment by nishitlangaliya** [comment]
@kloon ,

Thanks for reviewing my PR and provide your suggestions, I am doing changes as per your suggestions and will push the code into the same branch.

Thanks,

**Comment by nishitlangaliya** [comment]
@kloon ,
As per your suggestions I have updated the changes and push the code into branch, https://github.com/woocommerce/woocommerce/pull/17202/commits/77fe555339314ba880405cfb3dd61d4c4e83a096 please review it and let me know if any changes required.

Thanks.

File: includes/admin/class-wc-admin-assets.php [hunk 5]
```diff
@@ -156,6 +156,7 @@ public function admin_scripts() {
 				'i18_sale_less_than_regular_error'  => __( 'Please enter in a value less than the regular price.', 'woocommerce' ),
 				'decimal_point'                     => $decimal,
 				'mon_decimal_point'                 => wc_get_price_decimal_separator(),
+				'hexa_color_error'					=> __( 'Invalid hexa color', 'woocommerce' ),
```
Comment (shivapoudel) [inline includes/admin/class-wc-admin-assets.php]: Invalide hex color :)

Comment (nishitlangaliya) [reply includes/admin/class-wc-admin-assets.php]: @shivapoudel ,

Thanks for your suggestions, as per your suggestions I have updated the changes please check for the same and let me know if any changes required.

Thanks.

**Review by kloon — APPROVED** [review]

**Comment by kloon** [comment]
@nishitlangaliya couple of other issues that needs attention here.

1. The tooltip overlaps the colour picker http://cld.wthms.co/9as8Hn I'd suggest you close the colour picker when you show the tooltip, open to other suggestions.
2. After an error I can still save the option as it with the incorrect value, you should not be able to save the settings if the value is incorrect.
3. When you have an error on an input below another and you bring focus to the new input opening the colour picker the tooltip is still showing and overlapping http://cld.wthms.co/oTgLF3

**Review by kloon — CHANGES_REQUESTED** [review]
See comment for changes needed.

**Comment by nishitlangaliya** [comment]
@kloon,

Thanks for your suggestions. I will do required changes and let you know for the same.

Thanks

**Comment by nishitlangaliya** [comment]
Hi @kloon ,

As per suggestions I have done some changes and some query.

1. The tooltip overlaps the colour picker http://cld.wthms.co/9as8Hn I'd suggest you close the colour picker....

**[ Done ref #77b8f8b ]**

2. After an error I can still save the option as it with the incorrect value, ....
**[ In this I see woocommerce settings but can not find a reference like stop submit the form if validation occurs so here I need suggestion. Can we replace the default color if the user enters invalid hex color? also, can we prompt javascript alert and stop submit the form ? ]**

3. When you have an error on an input below another and you bring focus to the new input...
[ Done ref #77b8f8b ]

4. When you type a code and then click outside the input it just shows the tooltip briefly....
**[ I see in price edit also wc_error_tip is fade away if we click the outside. so should we display the error tip even user clicks out side. ? ]**

Please review my comment when you get a chance and let me know your feedback.

Thanks.

**Comment by mikejolley** [comment]
Close;  but there are some weird issues when you click out of the field. I think this is an unrelated issue. I'm looking into it.

**Comment by mikejolley** [comment]
I just found out iris has it's own validation we can use - it adds a class. iris-error

**Comment by mikejolley** [comment]
See https://github.com/woocommerce/woocommerce/pull/17493 Looks like IRIS can handle the validation for us and gives us a much cleaner way of doing things.

**Pull request closed without merging**